### ISSUE TITLE
Is the packed attention_mask compatible with transformers model?

### ISSUE THREAD

**Issue by mxjmtxrm** [MeetKai/functionary#195]
In the forward of LlamaFlashAttention2 in transformers, the attention mask is processed to `indices,cu_seqlens and max_seqlen in batch`. (transformers/models/llama/modeling_llama.py line 64)
```
def _get_unpad_data(attention_mask):
    seqlens_in_batch = attention_mask.sum(dim=-1, dtype=torch.int32)
    indices = torch.nonzero(attention_mask.flatten(), as_tuple=False).flatten()
    max_seqlen_in_batch = seqlens_in_batch.max().item()
    cu_seqlens = F.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
    return (
        indices,
        cu_seqlens,
        max_seqlen_in_batch,
    )
```
If  packed attention_mask is `[[1,1,1,1,2,2,2,3,3,3,0,0,0],[1,1,1,1,1,1,1,1,2,2,2,3,0]]`.
the output of the above function is:
```
indices=tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 13, 14, 15, 16, 17, 18, 19, 20,
        21, 22, 23, 24]), 
cu_seqlens=tensor([ 0, 19, 36], dtype=torch.int32), 
max_seqlen_in_batch=19
```
It is not correct as the max_seqlen_in_batch is 13. I think the expected output is:
```
indices=tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 13, 14, 15, 16, 17, 18, 19, 20,
        21, 22, 23, 24]), 
cu_seqlens=tensor([ 0, 4, 7, 13, 21, 24], dtype=torch.int32), 
max_seqlen_in_batch=13
```
Am I right? Is there anything I missed or misunderstood?
The lora training command is:
```
export CUDA_VISIBLE_DEVICES=7 
# deepspeed functionary/train/train_lora.py \
python functionary/train/train_lora.py \
    --model_name_or_path /workspace/model/Llama-2-7b-chat-hf \
    --train_data_path /workspace/data/train.json \
    --eval_data_path /workspace/data/train.json \
    --q_lora False \
    --bf16 True \
    --output_dir ./output \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --gradient_accumulation_steps 1 \
    --eval_accumulation_steps 1 \
    --evaluation_strategy "steps" \
    --eval_steps 5 \
    --save_strategy "steps" \
    --save_steps 5 \
    --save_total_limit 3 \
    --logging_steps 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --tf32 True \
    --model_max_length 1024 \
    --gradient_checkpointing True \
    --report_to tensorboard \
    --packing 

```

Looking forward to your reply.

**Comment by musabgultekin** [comment]
@khai-meetkai CC'ing

**Comment by khai-meetkai** [comment]
Hi @mxjmtxrm, if we use default: `_get_unpad_data` it **only works for normal attention_mask with 0 and 1**. 
For **packed attention_mask**, for example: [[1,1,1,1,2,2,2,3,3,3,0,0,0],[1,1,1,1,1,1,1,1,2,2,2,3,0]] We have to use the monkey-patched version: https://github.com/MeetKai/functionary/blob/main/functionary/train/packing/monkey_patch_packing.py#L21
You can see I replaced default `_get_unpad_data` at :https://github.com/MeetKai/functionary/blob/main/functionary/train/packing/monkey_patch_packing.py#L126C5-L126C78
`transformers.models.llama.modeling_llama._get_unpad_data = get_unpad_data `

So if you use packing, please make sure that: `monkey_patch_packing_llama()` would be run first

**Comment by khai-meetkai** [comment]
@mxjmtxrm , By the way, when you use the monkey-patched: https://github.com/MeetKai/functionary/blob/main/functionary/train/packing/monkey_patch_packing.py#L21
The output would be: 
```
indices = (tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 13, 14, 15, 16, 17, 18, 19, 20,
         21, 22, 23, 24]),
 cu_seqlens  = tensor([ 0,  4,  7, 10, 18, 21, 22], dtype=torch.int32),
 8)
```

For **cu_seqlens**, we don't include padding part (with 0)

**Comment by mxjmtxrm** [comment]
Got it. Thanks a lot.
